registry/registry.go

```diff
@@ -179,77 +168,186 @@ func (r dockerRegistry) removeImagePath(ctx context.Context, path string) error
 }
 
 func (r *dockerRegistry) doRequest(ctx context.Context, method, path string, headers map[string]string) (resp *http.Response, err error) {
-	u, _ := url.Parse(r.server)
-	server := r.server
+	u, _ := url.Parse(r.registry)
+	server := r.registry
 	if u != nil && u.Host != "" {
 		server = u.Host
 	}
-	authHeaders := registryAuth(server)
 	if r.client == nil {
 		r.client, err = tsuruNet.WithProxyFromConfig(*tsuruNet.Dial15Full300ClientNoKeepAlive, server)
 		if err != nil {
 			return nil, err
 		}
 	}
-	for _, scheme := range []string{"https", "http"} {
-		endpoint := fmt.Sprintf("%s://%s%s", scheme, server, path)
-		var req *http.Request
-		req, err = http.NewRequest(method, endpoint, nil)
-		if err != nil {
-			return nil, err
-		}
-		if ctx != nil {
-			req = req.WithContext(ctx)
-		}
-		req.Header = http.Header{}
-		for k, v := range headers {
-			req.Header.Set(k, v)
-		}
-		for k, v := range authHeaders {
-			req.Header[k] = v
-		}
-		resp, err = r.client.Do(req)
-		if err != nil {
+	max_tries := 5
+request:
+	for i := 0; i < max_tries; i++ {
+		for _, scheme := range []string{"https", "http"} {
+			endpoint := fmt.Sprintf("%s://%s%s", scheme, server, path)
+			var req *http.Request
+			req, err = http.NewRequest(method, endpoint, nil)
+			if err != nil {
+				return nil, err
+			}
+			if ctx != nil {
+				req = req.WithContext(ctx)
+			}
+			req.Header = http.Header{}
+			for k, v := range headers {
+				req.Header.Set(k, v)
+			}
+			r.fillAuthCredentials(req)
+			resp, err = r.client.Do(req)
 			if _, ok := err.(net.Error); ok {
 				continue
 			}
```

Comment:
@morpheu you miss to close the Request with
```
defer req.Body.Close()
```

to simplify that you could extract the inner loop to a function

Reply:
Added missing `resp.Body.Close()` inside fetchAuthToken logic